### PR TITLE
docs: add environmental-variables guide

### LINKED ISSUE #852: Write a guide for using environment variables
Write a guide about environment variables in tailcall. Cover aspects such as -
- Why is it needed?
- Where can one use it?
- Security aspects of using env variables.
- Some best practices

Make sure the guide is free-flowing and engaging. 

**PS: DON'T CREATE LOW-EFFORT CHATGPT GENERATED CONTENT!**

### PR DESCRIPTION
**Pull request by rajdip-b**
**Summary:**  
This PR adds a file called `environmental-variables.md` into the `docs/guides` folder. This guide is aimed at providing the users a clear picture about the whereabouts of environmental variables and how to use them in Tailcall.

**Issue Reference(s):**  
Fixes #852 
/claim #852

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [x] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcall/tree/main/docs

### CODE REVIEW

File: docs/guides/environmental-variables.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+---
+title: CLI
+---
+
+Environmental variables are key-value pairs that are stored in our operating systems. Many of them come by default, and we can also create our own. They are used to store information that is used by our operating system or by other programs. For example, the `PATH` variable stores a list of directories that the operating system should look in when we run a command in the terminal. The `HOME` variable stores the path to our home directory.
```
Comment (tusharmath) [inline docs/guides/environmental-variables.md:5]:
```suggestion
Environment variables are key-value pairs that are stored in our operating systems. Many of them come by default, and we can also create our own. They are used to store information that is used by our operating system or by other programs. For example, the `PATH` variable stores a list of directories that the operating system should look at when we run a command in the terminal. The `HOME` variable stores the path to our home directory.
```

Comment (rajdip-b) [reply docs/guides/environmental-variables.md:5]: okay there are multiple occurences of "environmental", should i replace them all? also, the file name has "environmental"

File: docs/guides/environmental-variables.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+The same can be used for building softwares aswell! Configuration values are often stored in environmental variables.
+
+## Need of Environmental Variables
+
+Applications use multiple external tools, authentication methods, and numerous configurations. Hence, for proper functioning, we need our code to use these values properly.
+
+Consider a simple scenario of JWT authentication. Typically, when signing the tokens for our users, we will need the following configuration set:
```
Comment (tusharmath) [inline docs/guides/environmental-variables.md:13]: Link JWT, incase the read doesn't know about JWT.

File: docs/guides/environmental-variables.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
+---
+title: CLI
+---
+
+Environmental variables are key-value pairs that are stored in our operating systems. Many of them come by default, and we can also create our own. They are used to store information that is used by our operating system or by other programs. For example, the `PATH` variable stores a list of directories that the operating system should look in when we run a command in the terminal. The `HOME` variable stores the path to our home directory.
+
+The same can be used for building softwares aswell! Configuration values are often stored in environmental variables.
```
Comment (tusharmath) [inline docs/guides/environmental-variables.md:7]: Avoid `!` in documentation. Keep the tone neutral.

File: docs/guides/environmental-variables.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+## Need of Environmental Variables
+
+Applications use multiple external tools, authentication methods, and numerous configurations. Hence, for proper functioning, we need our code to use these values properly.
+
+Consider a simple scenario of JWT authentication. Typically, when signing the tokens for our users, we will need the following configuration set:
+
+- expiry time
+- secret key
+- issuer
```
Comment (tusharmath) [inline docs/guides/environmental-variables.md:17]: Lighty describe what each of them does.

File: docs/guides/environmental-variables.md (hunk 3 above)
Comment (rajdip-b) [reply docs/guides/environmental-variables.md:7]: hmmmm got that.

File: docs/guides/environmental-variables.md (hunk 4 above)
Comment (rajdip-b) [reply docs/guides/environmental-variables.md:17]: sure

File: docs/guides/environmental-variables.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+To go about this, there are broadly two ways:
+
+1. **Hardcode the values in our code** \
+   This is perhaps the easiest approach to solve the problem, but the most dangerous (and inefficient) one! Once you hardcode the values in your codebase, everyone who works on the code will be able to see your sensitive information, which poses a massive security issue. Also, if you want to change the values, you will have to change the code and redeploy the application, which is not ideal.
+
+2. **Store the values in environmental variables** \
+   This is the best approach to solve the problem. You can store the sensitive values in the OS of the server running your application, and during the runtime of your application, you can access these values from the OS. All programming languages have excellent support for this. This way, you can keep your sensitive information safe, and also change the values without having to redeploy the application.
```
Comment (tusharmath) [inline docs/guides/environmental-variables.md:25]: We will need to restart the tailcall server though.

Comment (rajdip-b) [reply docs/guides/environmental-variables.md:25]: ahhh, yes, that's a case i missed.

**Comment by rajdip-b** [comment]
Made the changes!

File: docs/guides/environmental-variables.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+## Environmental variables in Tailcall
+
+With Tailcall, you can seamlessly fit in environmental variables into your GraphQL schema. Tailcall supports this by providing a `env` [Context](context.md) variable. This Context is then shared across all the operators, which you will use to resolve values in your schema.
+
+For example, you can do the following:
+
+```graphql showLineNumbers
+type Query {
+  users: [User]! @http(baseUrl: "{{env.API_ENDPOINT}}", path: "/users")
```
Comment (tusharmath) [inline docs/guides/environmental-variables.md:35]: - Expand on how this is better than hard coding the baseURL. For eg: You could have different URLs in Staging and Production,  etc.

- Can add this info - Env Variable can be used in all the places wherever we use mustache templates (in better wording).

File: docs/guides/environmental-variables.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+Environmental variables help you in mitigating most of the security risks. While this is true, it is also necessary to understand that, it doesn't actually help you in removing the risks completely, since the values are still in plain text. Configurational values might not be as sensitive, but the secrets still do have a chance to be compromised.
+
+To make sure your secrets stay safe and secure all through, here are some tips and tricks that you might consider using for your application:
+
+- **Use a `.env` file** \
+  This is a common practice that is followed by most developers. You can create a `.env` file in the root directory of your project and store all your environmental variables in it. This file should not be committed to your version control system, and should be added to the `.gitignore` file. This way, you can make sure that your secrets are not exposed to the public. Additionally, you can also use a `.env.example` file to list out all the environmental variables that are required for your application to run. This way, you can make sure that all the developers who work on your project are aware of the variables that are required for the application to run.
```
Comment (tusharmath) [inline docs/guides/environmental-variables.md:48]: Firstly Tailcall can't read .env files just yet. Secondly the best practices could be more from tailcall's point of view. For when to use an env variable and when not to use an env variable.

**Review by tusharmath — CHANGES_REQUESTED** [review]

File: docs/guides/environmental-variables.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+- **Store secrets through the GUI** \
+  If you are using a cloud provider to deploy your application, you can use the GUI provided by the cloud provider to store your environmental variables. They generally provide an intuitive UI, and furthermore, if your application is containerized and scales up and down automatically, this becomes the default go-to option.
+
+- **Using external services** \
+  There are many providers out there, mostly open-sourced softwares, that lift off the burden of security from your shoulders and provide you with a secure way to store your secrets. You can use these services to store your secrets and then access them in your application. Some of the popular ones are [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/), [Hashicorp Vault](https://www.vaultproject.io/), [Keywhiz](https://square.github.io/keywhiz/), etc.
```
Comment (tusharmath) [inline docs/guides/environmental-variables.md:57]: This whole section is more about managing secrets and less about env variables or tailcall.

File: docs/guides/environmental-variables.md (hunk 7 above)
Comment (rajdip-b) [reply docs/guides/environmental-variables.md:48]: got it

File: docs/guides/environmental-variables.md (hunk 6 above)
Comment (rajdip-b) [reply docs/guides/environmental-variables.md:35]: sure will do

File: docs/guides/environmental-variables.md (hunk 8 above)
Comment (rajdip-b) [reply docs/guides/environmental-variables.md:57]: okay then, should i add more examples in here about where and how we can use environment variables? or do you have something else in mind?

File: docs/guides/environmental-variables.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+- **Secret key**: The secret key that will be used to encrypt the token
+- **Issuer**: The name of the issuer of the token. This is generally the organization's name.
+
+To go about this, there are broadly two ways:
+
+1. **Hardcode the values in our code** \
+   This is perhaps the easiest approach to solve the problem, but the most dangerous (and inefficient) one. Once you hardcode the values in your codebase, everyone who works on the code will be able to see your sensitive information, which poses a massive security issue. Also, if you want to change the values, you will have to change the code and redeploy the application, which is not ideal.
+
+2. **Store the values in environmental variables** \
+   This is the best approach to solve the problem. You can store the sensitive values in the OS of the server running your application, and during the runtime of your application, you can access these values from the OS. All programming languages have excellent support for this. This way, you can keep your sensitive information safe, and also change the values without having to change anything in your code.
```
Comment (ologbonowiwi) [inline docs/guides/environmental-variables.md:25]: This is not particularly necessary as it does not relate to tailcall directly.

```suggestion
```

File: docs/guides/environmental-variables.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
+---
+title: CLI
+---
+
+Environment variables are key-value pairs that are stored in our operating systems. Many of them come by default, and we can also create our own. They are used to store information that is used by our operating system or by other programs. For example, the `PATH` variable stores a list of directories that the operating system should look in when we run a command in the terminal. The `HOME` variable stores the path to our home directory.
+
+The same can be used for building softwares aswell. Configuration values are often stored in environmental variables.
```
Comment (ologbonowiwi) [inline docs/guides/environmental-variables.md:7]:
```suggestion
The same can be used for building software as well. Configuration values are often stored in environmental variables.
```

**Review by ologbonowiwi — CHANGES_REQUESTED** [review]

File: docs/guides/environmental-variables.md (hunk 9 above)
Comment (rajdip-b) [reply docs/guides/environmental-variables.md:25]: @tusharmath asked to give some context in the issue aswell, so thought mught include it. I can remove it if you guys want me to!

File: docs/guides/environmental-variables.md (hunk 10 above)
Comment (rajdip-b) [reply docs/guides/environmental-variables.md:7]: fixed in upcoming PR, awaiting some confirmations rn.

File: docs/guides/environmental-variables.md (hunk 9 above)
Comment (ologbonowiwi) [reply docs/guides/environmental-variables.md:25]: better hear @tusharmath thoughts here then; if we create more examples as mentioned on [this comment](https://github.com/tailcallhq/tailcall/pull/855#discussion_r1438870730)
> Expand on how this is better than hard coding the baseURL. For eg: You could have different URLs in Staging and Production, etc.

IMO we should be good to delete this part of it.

Comment (rajdip-b) [reply docs/guides/environmental-variables.md:25]: Yeap I am working on adding more examples and places where we can use env vars.

**Comment by rajdip-b** [comment]
@tusharmath made the requested changes!

File: docs/guides/environmental-variables.md (hunk 9 above)
Comment (tusharmath) [reply docs/guides/environmental-variables.md]: @ologbonowiwi Thanks a ton for reviewing! This content will go into the "guides" section and not the "docs" section. Guides should go a little beyond just tailcall. We will be doing more of this going forward. In this guide, we are targeting newcomers who haven't used environment variables before.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**